word2gauss/embeddings.pyx

```diff
@@ -0,0 +1,944 @@
+        cdef np.ndarray[DTYPE_t, ndim=2, mode='c'] _sigma
+        cdef np.ndarray[DTYPE_t, ndim=1, mode='c'] _acc_grad_mu
+        cdef np.ndarray[DTYPE_t, ndim=1, mode='c'] _acc_grad_sigma
+
+        if covariance_type == 'spherical':
+            self.covariance_type = SPHERICAL
+        else:
+            raise ValueError
+    
+        self.energy_type = energy_type
+        if energy_type == 'KL':
+            self.energy_func = <energy_t>kl_energy
+            self.gradient_func = <gradient_t>kl_gradient
+        elif energy_type == 'IP':
+            self.energy_func = <energy_t>ip_energy
+            self.gradient_func = <gradient_t>ip_gradient
+        else:
+            raise ValueError
+
+        self.N = N 
+        self.K = size
+        self.sigma_min = sigma_min
+        self.sigma_max = sigma_max
+        self.mu_max = mu_max
+        self.eta = eta
+        self.Closs = Closs
+
+        # Initialize parameters 
+        if mu is None:
+            _mu = init_params['mu0'] * np.ascontiguousarray(
+                np.random.randn(self.N, self.K).astype(DTYPE))
+        else:
+            assert mu.shape[0] == N
+            assert mu.shape[1] == self.K
+            _mu = mu
+        self.mu = _mu
+
+        if sigma is None:
+            if self.covariance_type == SPHERICAL:
+                s = np.random.randn(self.N, 1).astype(DTYPE)
+            s *= init_params['sigma_std0']
+            s += init_params['sigma_mean0']
+            _sigma = np.ascontiguousarray(
+                np.maximum(sigma_min, np.minimum(s, sigma_max)))
+        else:
+            assert sigma.shape[0] == N
+            if self.covariance_type == SPHERICAL:
+                assert sigma.shape[1] == 1
+            _sigma = sigma
+        self.sigma = _sigma
+
+        assert _mu.flags['C_CONTIGUOUS']
+        assert _sigma.flags['C_CONTIGUOUS']
+        self.mu_ptr = &_mu[0, 0]
+        self.sigma_ptr = &_sigma[0, 0]
+
+        # working space for accumulated gradients
+        _acc_grad_mu = np.ascontiguousarray(np.zeros(N, dtype=DTYPE))
+        assert _acc_grad_mu.flags['C_CONTIGUOUS']
+        self.acc_grad_mu = _acc_grad_mu
+        self.acc_grad_mu_ptr = &_acc_grad_mu[0]
+
+        if self.covariance_type == SPHERICAL:
+            _acc_grad_sigma = np.ascontiguousarray(np.zeros(N, dtype=DTYPE))
+        assert _acc_grad_sigma.flags['C_CONTIGUOUS']
+        self.acc_grad_sigma = _acc_grad_sigma
+        self.acc_grad_sigma_ptr = &_acc_grad_sigma[0]
+
+    def save(self, fname, vocab=None, full=True):
+        '''
+        Writes a gzipped text file of the model in word2vec text format
+
+        word_or_id1 0.5 -0.2 1.0 ...
+        word_or_id2 ...
+
+        This class doesn't have knowledge of id -> word mapping, so
+        it can be passed in as a callable vocab(id) return the word string
+
+        if full=True, then writes out the full model.  It is a tar.gz
+        file the word vectors and Sigma files, in addition
```

Comment:
Missing a word here? 'file the word vectors...'  -> 'file containing the word vectors...'